File: python/leetcode039.py
class Solution(object):

    def combinationSum(self, candidates, target):
        candidates.sort()
        dp = [[] for _ in range(target + 1)]
        dp[0].append([])
        for i in range(1, target + 1):
            for j in range(len(candidates)):
                if candidates[j] > i:
                    break
                for k in range(len(dp[i - candidates[j]])):
                    temp = dp[i - candidates[j]][k][:]
                    if len(temp) > 0 and temp[-1] > candidates[j]:
                        continue
                    temp.append(candidates[j])
                    dp[i].append(temp)
        return dp[target]


if __name__ == '__main__':
    s = Solution()
    print(s.combinationSum([8,7,4,3], 200))

File: leetcodec++/leetcode039.cpp
//
// Created by 黄敏辉 on 2019-07-11.
//

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        sort(candidates.begin(), candidates.end());
        vector<int> comb;
        helper(candidates, target, res, comb, 0);
        return res;
    }

    void helper(vector<int>& candidates, int target, vector<vector<int>>& res, vector<int> comb, int begin){
        if (target == 0){
            res.push_back(comb);
            return;
        }

        for(int i = begin; i != candidates.size() && candidates[i] <= target; i++){
            comb.push_back(candidates[i]);
            helper(candidates, target - candidates[i], res, comb, i);
            comb.pop_back();
        }
    }
};